hail/src/main/scala/is/hail/expr/ir/Emit.scala

```diff
@@ -3684,32 +3625,96 @@ class Emit[C](val ctx: EmitContext, val cb: EmitClassBuilder[C]) {
       cb.memoize(cb.invokeCode[Boolean](sort, cb.this_, region, l, r))
   }
 
-  def emitLet[A](
-    emitI: (IR, EmitCodeBuilder, EmitEnv) => IEmitCode,
-    emitBody: (IR, EmitCodeBuilder, EmitEnv) => A,
+  def emitLetBindings(
+    emitI: (IR, EmitCodeBuilder, EmitEnv, Value[Region]) => IEmitCode,
+    emitVoid: (IR, EmitCodeBuilder, EmitEnv, Value[Region]) => Unit,
   )(
     let: Let,
     cb: EmitCodeBuilder,
     env: EmitEnv,
-  ): A = {
+    r: Value[Region],
+  ): EmitEnv = {
     val uses: mutable.Set[String] =
       ctx.usesAndDefs.uses.get(let) match {
         case Some(refs) => refs.map(_.t.name)
         case None => mutable.Set.empty
       }
 
-    emitBody(
-      let.body,
-      cb,
-      let.bindings.foldLeft(env) { case (newEnv, (name, ir)) =>
-        if (!uses.contains(name)) newEnv
-        else {
-          val value = emitI(ir, cb, newEnv)
+    def emitChunk(cb: EmitCodeBuilder, bindings: Seq[(String, IR)], env: EmitEnv, r: Value[Region])
+      : EmitEnv =
+      bindings.foldLeft(env) { case (newEnv, (name, ir)) =>
+        if (ir.typ == TVoid) {
+          emitVoid(ir, cb, newEnv, r)
+          newEnv
+        } else if (IsPure(ir) && !uses.contains(name)) {
+          newEnv
+        } else {
+          val value = emitI(ir, cb, newEnv, r)
           val memo = cb.memoizeMaybeStreamValue(value, s"let_$name")
           newEnv.bind(name, memo)
         }
-      },
-    )
+      }
+
+    @tailrec def go(
+      env: EmitEnv,
+      chunkStart: Int,
+      pos: Int,
```

Comment:
What's `pos` for? There seems to be three ints we're tracking at once when wouldn't two be enough to define the window of the bindings to emit in one group?

Reply:
Added explanatory comments.